### DISCUSSION TITLE
请问一下在prefill计算时，ProcessKV和FlattenKV的作用分别是什么？

### DISCUSSION THREAD

**Discussion by sitabulaixizawaluduo** [InternLM/lmdeploy#2063] [Q&A]
是不是可以理解ProcessKV是把经过qkv_proj计算的k v存到block里面，FlattenKV是把block kv cache转为整块的kv cache参与后续的attention计算？

**Comment by lzhangzz** [comment]
是的

**Reply by sitabulaixizawaluduo** [reply]
我看到ProcessKV和FlattenKV分别做了一次rope，FlattenKV中把kv cache从block中取出来时为什么还要再做一次rope呢，这个操作是针对什么场景的？

**Reply by snippetzero** [reply]
同样好奇

**Reply by lzhangzz** [reply]
这两个地方不会同时生效
